packages/cloud_firestore/android/src/main/java/io/flutter/plugins/firebase/cloudfirestore/CloudFirestorePlugin.java

```diff
@@ -208,6 +233,127 @@ public void onFailure(@NonNull Exception e) {
+
+              // Start operations on dart side.
+              channel.invokeMethod("DoTransaction", arguments, new Result() {
+                @Override
+                public void success(Object o) {
+                  transactionTCS.setResult((Map<String, Object>) o);
+                }
+
+                @Override
+                public void error(String s, String s1, Object o) {
+                  result.error(s, s1, o);
+                  transactionTCS.setResult(null);
+                }
+
+                @Override
+                public void notImplemented() {
+                  result.error("DoTransaction not implemented", null, null);
+                  transactionTCS.setResult(null);
+                }
+              });
+
+              // wait till transaction is complete.
+              try {
+                Map<String, Object> transactionResult = Tasks.await(
+                        transactionTCSTask,
+                        getTransactionTimeout(arguments),
+                        TimeUnit.MILLISECONDS);
+                result.success(transactionResult);
+              } catch (Exception e) {
+                result.error("Error performing transaction", e.getMessage(), null);
+              }
+              return null;
+            }
+          });
+          break;
+        }
+      case "Transaction#get":
+        {
+          final Map<String, Object> arguments = call.arguments();
+          final Transaction transaction = getTransaction(arguments);
+          new AsyncTask<Void, Void, Void>() {
+            @Override
+            protected Void doInBackground(Void... voids) {
+              try {
+                DocumentSnapshot documentSnapshot = transaction.get(getDocumentReference(arguments));
+                Map<String, Object> snapshotMap = new HashMap<>();
+                snapshotMap.put("path", documentSnapshot.getReference().getPath());
+                if (documentSnapshot.exists()) {
+                  snapshotMap.put("data", documentSnapshot.getData());
+                } else {
+                  snapshotMap.put("data", null);
+                }
+                result.success(snapshotMap);
+              } catch (FirebaseFirestoreException e) {
+                result.error("Error performing Transaction#get", e.getMessage(), null);
+              }
+              return null;
+            }
+          }.execute();
+          break;
+        }
+      case "Transaction#update":
+        {
+          final Map<String, Object> arguments = call.arguments();
+          final Transaction transaction = getTransaction(arguments);
+          new AsyncTask<Void, Void, Void>() {
+            @Override
+            protected Void doInBackground(Void... voids) {
+              Map<String, Object> data = (Map<String, Object>) arguments.get("data");
+              transaction.update(getDocumentReference(arguments), data);
+              result.success(null);
```

Comment:
As you pointed out, there's probably a race condition where it's possible this code could run after the FlutterActivity is torn down. It might throw an exception that could show up in crash reports; I'm not sure how much we care about this, but it might be worth simulating that race with a sleep just to see what happens.